### ISSUE TITLE
在dtk上 lstm与bn的前向都不支持fp16

### ISSUE THREAD

**Issue by catsled** [PaddlePaddle/Paddle#73457]
## bn
```
import paddle
import numpy as np

# 1. 明确设置默认 dtype 为 float16（影响后续构建）
paddle.set_default_dtype("float16")

# 2. 构造 BatchNorm 层
bn = paddle.nn.BatchNorm2D(num_features=4)

# 3. 查看参数信息（应该是 float16）
print("BatchNorm 参数:")
for name, param in bn.named_parameters():
    print(f"{name}: dtype={param.dtype}, shape={param.shape}")

# 4. 生成 float16 输入数据
x = paddle.randn([2, 4, 8, 8], dtype='float16')

# 5. 前向推理
bn.eval()
out = bn(x)
print("前向输出 dtype:", out.dtype)
```
结果
BatchNorm 参数:
weight: dtype=paddle.float32, shape=[4]
bias: dtype=paddle.float32, shape=[4]
_mean: dtype=paddle.float32, shape=[4]
_variance: dtype=paddle.float32, shape=[4]
前向输出 dtype: paddle.float16

## lstm

```
import numpy as np

paddle.set_device('gpu')
paddle.set_default_dtype('float16')  # 默认尝试 float16

# 1. 构建 float16 输入
x_fp16 = paddle.to_tensor(np.random.randn(5, 2, 16).astype('float16'))  # [seq_len, batch, input_size]
x_fp32 = paddle.to_tensor(np.random.randn(5, 2, 16).astype('float32'))
print("Input dtype:", x_fp16.dtype)

# 2. 构造 LSTM 层
try:
    lstm = paddle.nn.LSTM(input_size=16, hidden_size=32, num_layers=1)
    print("LSTM 参数 dtype：")
    for name, param in lstm.named_parameters():
        print(f"{name}: {param.dtype}")
except Exception as e:
    print("LSTM 初始化异常:", e)

# 3. 尝试前向
try:
    lstm.eval()
    out, (h, c) = lstm(x_fp16)  # 如果 float16 不支持，会在这里报错
    print("输出 dtype:", out.dtype)
except Exception as e:
    print("LSTM 前向计算失败，错误：", e)
```
weight_ih_l0: paddle.float32
weight_hh_l0: paddle.float32
bias_ih_l0: paddle.float32
bias_hh_l0: paddle.float32
LSTM 前向计算失败，错误： (NotFound) The kernel with key (GPU, Undefined(AnyLayout), float16) of kernel `rnn` is not registered and fail to fallback to CPU one. Selected wrong DataType `float16`. Paddle support following DataTypes: float32.
  [Hint: Expected kernel_iter != iter->second.end(), but received kernel_iter == iter->second.end().] (at /home/bw_ut_pp/3.0/2504/paddle/paddle/phi/core/kernel_factory.cc:380)
